src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java

```diff
@@ -209,34 +225,114 @@ private void loadConfig(Map<String, Object> config, AmazonCloudWatch client) {
           }
+
+          if (yamlMetricRule.containsKey("aws_tag_select")) {
+            Map<String, Object> yamlAwsTagSelect = (Map<String, Object>) yamlMetricRule.get("aws_tag_select");
+            if (!yamlAwsTagSelect.containsKey("resource_type_selection") || !yamlAwsTagSelect.containsKey("resource_id_dimension") ||
+              !yamlAwsTagSelect.containsKey("tag_selections")) {
+              throw new IllegalArgumentException("Must provide resource_type_selection, resource_id_dimension and tag_selections");
+            }
+            AWSTagSelect awsTagSelect = new AWSTagSelect();
+            rule.awsTagSelect = awsTagSelect;
+
+            awsTagSelect.resourceTypeSelection = (String)yamlAwsTagSelect.get("resource_type_selection");
+            awsTagSelect.resourceIdDimension = (String)yamlAwsTagSelect.get("resource_id_dimension");
+            awsTagSelect.tagSelections = (Map<String, List<String>>)yamlAwsTagSelect.get("tag_selections");
+          }
         }
 
-        loadConfig(rules, client);
+        loadConfig(rules, cloudWatchClient, taggingClient);
     }
 
-    private void loadConfig(ArrayList<MetricRule> rules, AmazonCloudWatch client) {
+    private void loadConfig(ArrayList<MetricRule> rules, AmazonCloudWatch cloudWatchClient, AWSResourceGroupsTaggingAPI taggingClient) {
         synchronized (activeConfig) {
-            activeConfig.client = client;
+            activeConfig.cloudWatchClient = cloudWatchClient;
+            activeConfig.taggingClient = taggingClient;
             activeConfig.rules = rules;
         }
     }
 
+    private AWSCredentialsProvider getRoleCredentialProvider(Map<String, Object> config) {
+        STSAssumeRoleSessionCredentialsProvider credentialsProvider = new STSAssumeRoleSessionCredentialsProvider.Builder(
+                (String) config.get("role_arn"),
+                "cloudwatch_exporter"
+              ).build();
+        return credentialsProvider;
+    }
+    
+    private Region getRegion(Map<String, Object> config) {
+        Region region = RegionUtils.getRegion((String) config.get("region"));
+        if (region == null) {
+          region = Regions.getCurrentRegion();
+          if (region == null) {
+            throw new IllegalArgumentException("No region provided and EC2 metadata failed");
+          }
+        }
+        return region;
+    }
+    
     public String getMonitoringEndpoint(Region region) {
-      return "https://" + region.getServiceEndpoint("monitoring");
+        return "https://" + region.getServiceEndpoint("monitoring");
     }
 
-    private List<List<Dimension>> getDimensions(MetricRule rule, AmazonCloudWatch client) {
+    private List<ResourceTagMapping> getResourceTagMappings(MetricRule rule, AWSResourceGroupsTaggingAPI taggingClient) {
+	List<ResourceTagMapping> resourceTagMappings = new ArrayList<ResourceTagMapping>();
+	if(rule.awsTagSelect == null) {
+            return resourceTagMappings;
```

Comment:
If the list is not intended to be modified return Collections.<ResourceTagMapping>emptyList() instead. It would not create an instance at each call.

Reply:
```suggestion
            return Collections.<ResourceTagMapping>emptyList();
```

Reply:
@vincentcreusotfx ok, good one, fixed (even though I did not take your code suggestion, as the generic is implicitly detected by the compiler here!)